src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,667 @@
+   *
+   * @return the underlying RestHighLevelClient
+   */
+  public RestHighLevelClient client() {
+    return client;
+  }
+
+  /**
+   * Closes the ElasticsearchClient.
+   *
+   * @throws ConnectException if all of the records fail to flush before the timeout.
+   */
+  public void close() {
+    try {
+      if (!bulkProcessor.awaitClose(config.flushTimeoutMs(), TimeUnit.MILLISECONDS)) {
+        throw new ConnectException("Failed to process all outstanding requests in time.");
+      }
+    } catch (InterruptedException e) {
+      log.error("Bulk processor close was interrupted.", e);
+    }
+
+    try {
+      if (connectionReaper != null) {
+        connectionReaper.shutdown();
+        connectionReaper.join();
+        connectionReaper = null;
+      }
+    } catch (InterruptedException e) {
+      log.error("Interrupted while closing connection evicting thread.");
+    }
+
+    try {
+      client.close();
+    } catch (IOException e) {
+      log.error("Failed to close Elasticsearch client.", e);
+    }
+  }
+
+  /**
+   * Creates an index. Will not recreate the index if it already exists.
+   *
+   * @param index the index to create
+   * @return true if the index was created, false if it already exists
+   */
+  public boolean createIndex(String index) {
+    if (indexExists(index)) {
+      return false;
+    }
+
+    CreateIndexRequest request = new CreateIndexRequest(index);
+    return callWithRetries(
+        "create index " + index,
+        () -> {
+          try {
+            client.indices().create(request, RequestOptions.DEFAULT);
+          } catch (ElasticsearchStatusException | IOException e) {
+            if (!e.getMessage().contains(RESOURCE_ALREADY_EXISTS_EXCEPTION)) {
+              throw e;
+            }
+            return false;
+          }
+          return true;
+        }
+    );
+  }
+
+  /**
+   * Creates a mapping for the given index and schema.
+   *
+   * @param index the index to create the mapping for
+   * @param schema the schema to map
+   */
+  public void createMapping(String index, Schema schema) {
+    PutMappingRequest request = new PutMappingRequest(index).source(Mapping.buildMapping(schema));
+    callWithRetries(
+        String.format("create mapping for index %s with schema %s", index, schema),
+        () -> client.indices().putMapping(request, RequestOptions.DEFAULT)
+    );
+  }
+
+  /**
+   * Flushes any buffered records.
+   */
+  public void flush() {
+    bulkProcessor.flush();
+  }
+
+  /**
+   * Checks whether the index already has a mapping or not.
+   * @param index the index to check
+   * @return true if a mapping exists, false if it does not
+   */
+  public boolean hasMapping(String index) {
+    MappingMetaData mapping = mapping(index);
+    return mapping != null && !mapping.sourceAsMap().isEmpty();
+  }
+
+  /**
+   * Buffers a record to index. Will ensure that there are no concurrent requests for the same
+   * document id.
```

Comment:
IIUC, the latter statement is valid for a case when DLQ is enabled or `key.ignore=false`. Maybe update the Javadoc and mention, why we are not ensuring the absence of concurrent requests for some cases.